### PR TITLE
Introduce getJavaClass() and getJavaMethod() in MethodSource

### PR DESCRIPTION
**Pull request by cberes**
Issue: #2294

## Overview

I saw there was some discussion about the issue reporter's working on a PR, but it had been almost a month so I hope I'm not stepping on any toes!

I added a getJavaMethods() method as well, because I was unsure what to do about overloaded methods. I see there is a methodParameterTypes field, which I imagine I could use to find the exact specified method. But I'm unsure about the format of the field and how consistent that format would be. Also does a null methodParameterTypes field mean there are no arguments, or just that no arguments were specified?

Also I made getJavaMethod() throw exceptions if there are 0 or >1 methods with the given name. Unsure if this is the desired behavior.

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [ ] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [ ] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [ ] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

File: junit-platform-engine/src/main/java/org/junit/platform/engine/support/descriptor/MethodSource.java [hunk 1]
```diff
@@ -167,6 +175,67 @@ public boolean equals(Object o) {
...
+	 * the method's name and throws a {@link PreconditionViolationException} if
+	 * the methods cannot be loaded.
+	 *
+	 * <p>An empty list will be returned if there are no matching methods.
+	 *
+	 * @since 5.7
+	 * @see #getMethodName()
+	 * @see #getJavaMethod()
+	 */
+	public List<Method> getJavaMethods() {
```
Comment (sbrannen) [inline junit-platform-engine/src/main/java/org/junit/platform/engine/support/descriptor/MethodSource.java:210]: As mentioned in https://github.com/junit-team/junit5/issues/2294#issuecomment-629152991, we are only adding `getJavaClass()` and `getJavaMethod()`.  Thus, please remove the `getJavaMethods()` method.

File: junit-platform-engine/src/main/java/org/junit/platform/engine/support/descriptor/MethodSource.java [hunk 2]
```diff
@@ -167,6 +175,67 @@ public boolean equals(Object o) {
...
+	 * <p>If the {@link Method} was not provided, but only the name, this method
+	 * attempts to lazily load the {@link Method} based on its name and throws a
+	 * {@link PreconditionViolationException} if the method cannot be loaded or
+	 * if the method has overloads.
+	 *
+	 * @since 5.7
+	 * @see #getMethodName()
+	 * @see #getJavaMethods()
+	 */
+	public Method getJavaMethod() {
```
Comment (sbrannen) [inline junit-platform-engine/src/main/java/org/junit/platform/engine/support/descriptor/MethodSource.java:230]: Please implement this method analogous to `org.junit.platform.engine.discovery.MethodSelector.getJavaMethod()`.

File: junit-platform-engine/src/main/java/org/junit/platform/engine/support/descriptor/MethodSource.java [hunk 3]
```diff
@@ -167,6 +175,67 @@ public boolean equals(Object o) {
...
+	 * Get the {@linkplain Class Java class} of this source.
+	 *
+	 * <p>If the {@link Class} was not provided, but only the name, this method
+	 * attempts to lazily load the {@link Class} based on its name and throws a
+	 * {@link PreconditionViolationException} if the class cannot be loaded.
+	 *
+	 * @since 5.7
+	 * @see #getClassName()
+	 */
+	public final Class<?> getJavaClass() {
```
Comment (sbrannen) [inline junit-platform-engine/src/main/java/org/junit/platform/engine/support/descriptor/MethodSource.java:188]: Please add missing `@API` declaration.

File: junit-platform-engine/src/main/java/org/junit/platform/engine/support/descriptor/MethodSource.java (hunk 2 above)
Comment (sbrannen) [inline junit-platform-engine/src/main/java/org/junit/platform/engine/support/descriptor/MethodSource.java:230]: Please add missing `@API` declaration.

**Comment by sbrannen** [comment]
Once you have made the requested changes and updated the tests accordingly, please remember that you need to document the new features as listed in the deliverables.

- Change is documented in the User Guide and Release Notes

**Comment by cberes** [comment]
Thank you for the feedback.

I added a comment about the new methods to the release notes, though I'm unsure of what I should add to the user guide. There's a blurb about MethodSource in [URI Test Sources for Dynamic Tests](https://junit.org/junit5/docs/snapshot/user-guide/#writing-tests-dynamic-tests-uri-test-source), but that seems like an inappropriate place to mention these new methods. I looked in the user guide for some related classes such as ClassSource and TestIdentifier but I didn't find them.

File: junit-platform-engine/src/main/java/org/junit/platform/engine/support/descriptor/MethodSource.java (hunk 1 above)
Comment (cberes) [reply junit-platform-engine/src/main/java/org/junit/platform/engine/support/descriptor/MethodSource.java]: sounds good

File: junit-platform-engine/src/main/java/org/junit/platform/engine/support/descriptor/MethodSource.java (hunk 2 above)
Comment (cberes) [reply junit-platform-engine/src/main/java/org/junit/platform/engine/support/descriptor/MethodSource.java:208]: ok

File: junit-platform-engine/src/main/java/org/junit/platform/engine/support/descriptor/MethodSource.java (hunk 3 above)
Comment (cberes) [reply junit-platform-engine/src/main/java/org/junit/platform/engine/support/descriptor/MethodSource.java:188]: ok. I assume STABLE is the correct value

File: junit-platform-engine/src/main/java/org/junit/platform/engine/support/descriptor/MethodSource.java (hunk 2 above)
Comment (cberes) [reply junit-platform-engine/src/main/java/org/junit/platform/engine/support/descriptor/MethodSource.java:208]: ok

File: junit-platform-engine/src/main/java/org/junit/platform/engine/support/descriptor/MethodSource.java (hunk 3 above)
Comment (sbrannen) [reply junit-platform-engine/src/main/java/org/junit/platform/engine/support/descriptor/MethodSource.java:188]: Yes, STABLE should be fine.

**Comment by sbrannen** [comment]
> Thank you for the feedback.

You're welcome.

> I added a comment about the new methods to the release notes, 

Thanks

> though I'm unsure of what I should add to the user guide. There's a blurb about MethodSource in [URI Test Sources for Dynamic Tests](https://junit.org/junit5/docs/snapshot/user-guide/#writing-tests-dynamic-tests-uri-test-source), but that seems like an inappropriate place to mention these new methods. 

Right. That would be an inappropriate place.

> I looked in the user guide for some related classes such as ClassSource and TestIdentifier but I didn't find them.

Indeed. These features aren't actually documented in the User Guide. So no need to document these new methods in `MethodSource`.

File: documentation/src/docs/asciidoc/release-notes/release-notes-5.7.0-M2.adoc [hunk 4]
```diff
@@ -40,6 +40,7 @@ on GitHub.
 * Support `FilePosition` in `FileSelector` and `ClasspathResourceSelector`.
 * Custom `PostDiscoveryFilter` implementations can now be registered via Java’s
   `ServiceLoader` mechanism.
+* New `getJavaClass` and `getJavaMethod` methods in class `MethodSource`
```
Comment (sbrannen) [inline documentation/src/docs/asciidoc/release-notes/release-notes-5.7.0-M2.adoc:43]:
```suggestion
* New `getJavaClass()` and `getJavaMethod()` methods in
   `org.junit.platform.engine.support.descriptor.MethodSource`.
```

**Comment by sbrannen** [comment]
This has been merged into `master` in a91115b6edc94bb02a1f0863d4c11ddb8536777f and polished in cb5e5e9439da6381d6f2743adb23fabd387c5ac5.

Thanks

**Pull request closed without merging**